### PR TITLE
Cross Border Class Tax Calculation

### PR DESCRIPTION
**Pull request by LordZardeck**

### CODE REVIEW

File: Plugin/Model/ResourceModel/Quote/ExtensionAttributesPersistencePlugin.php [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    public function __construct(
+        JoinProcessorHelper $joinProcessorHelper,
+        ExtensionAttributesFactory $extensionAttributesFactory
+    )
+    {
+        $this->joinProcessorHelper = $joinProcessorHelper;
+        $this->extensionAttributesFactory = $extensionAttributesFactory;
+    }
+
+    protected function getJoinDirectivesForType($extensibleEntityClass)
```
Comment (erikhansen) [inline Plugin/Model/ResourceModel/Quote/ExtensionAttributesPersistencePlugin.php]: Add DocBlock, including a comment(s) describing what this does

File: Plugin/Model/ResourceModel/ExtensionAttributesPersistencePlugin.php [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    public function __construct(
+        ExtensionAttributesFactory $extensionAttributesFactory,
+        ExtensionAttributeMerger $extensionAttributeMerger
+    )
+    {
+        $this->extensionAttributesFactory = $extensionAttributesFactory;
+        $this->extensionAttributeMerger = $extensionAttributeMerger;
+    }
+
+    protected function getJoinDirectivesForType($extensibleEntityClass)
```
Comment (erikhansen) [inline Plugin/Model/ResourceModel/ExtensionAttributesPersistencePlugin.php]: Add DocBlock, including a comment(s) describing what this does

File: Plugin/Model/ResourceModel/Quote/ExtensionAttributesPersistencePlugin.php (hunk 1 above)
Comment (erikhansen) [reply Plugin/Model/ResourceModel/Quote/ExtensionAttributesPersistencePlugin.php]: Disregard this comment. I'm going to add all comments to the PR, not the individual commits.

File: Plugin/Model/ResourceModel/ExtensionAttributesPersistencePlugin.php (hunk 2 above)
Comment (erikhansen) [inline Plugin/Model/ResourceModel/ExtensionAttributesPersistencePlugin.php]: Add DocBlock, including a comment(s) describing what this does

File: Plugin/Model/ResourceModel/ExtensionAttributesPersistencePlugin.php [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    /**
+     * @param AbstractDb    $subject
+     * @param callable      $proceed
+     * @param AbstractModel $object
+     *
+     * @return mixed
+     */
+    public function aroundSave(AbstractDb $subject, callable $proceed, AbstractModel $object)
```
Comment (erikhansen) [inline Plugin/Model/ResourceModel/ExtensionAttributesPersistencePlugin.php]: This code needs more comments, explaining what this method does

File: Plugin/Model/ResourceModel/ExtensionAttributesPersistencePlugin.php [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+                $tableName,
+                array_merge(...$tableData[$tableName]),
+                array_merge(...$tableFields[$tableName])
+            );
+        }
+
+        return $subject;
+    }
+
+    public function aroundLoad(AbstractDb $subject, callable $proceed, AbstractModel $object, $value, $field = null)
```
Comment (erikhansen) [inline Plugin/Model/ResourceModel/ExtensionAttributesPersistencePlugin.php]: Add DocBlock, including a comment(s) describing what this does

File: Plugin/Data/Collection/ExtensionAttributeCollectionLoader.php [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+    /**
+     * @param \Magento\Framework\Api\ExtensionAttribute\JoinProcessorInterface $joinProcessor
+     */
+    public function __construct(\Magento\Framework\Api\ExtensionAttribute\JoinProcessorInterface $joinProcessor)
+    {
+        $this->joinProcessor = $joinProcessor;
+    }
+
+    public function beforeLoad(
```
Comment (erikhansen) [inline Plugin/Data/Collection/ExtensionAttributeCollectionLoader.php]: Add DocBlock

File: Helper/ExtensionAttributeMerger.php [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    public function __construct(
+        JoinProcessorHelper $joinProcessorHelper,
+        \Magento\Framework\Api\ExtensionAttributesFactory $extensionAttributesFactory
+    )
+    {
+        $this->extensionAttributesFactory = $extensionAttributesFactory;
+        $this->joinProcessorHelper = $joinProcessorHelper;
+    }
+
+    protected function getExtensionAttributeMethodName($key)
```
Comment (erikhansen) [inline Helper/ExtensionAttributeMerger.php]: Add DocBlocks to this and the methods below. If you copied this code from core methods, add a `@see` reference

File: Plugin/Model/Convert/OrderPlugin.php [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+    /**
+     * @param \ClassyLlama\AvaTax\Helper\ExtensionAttributeMerger $extensionAttributeMerger
+     */
+    public function __construct(\ClassyLlama\AvaTax\Helper\ExtensionAttributeMerger $extensionAttributeMerger)
+    {
+        $this->extensionAttributeMerger = $extensionAttributeMerger;
+    }
+
+    public function aroundItemToInvoiceItem(
```
Comment (erikhansen) [inline Plugin/Model/Convert/OrderPlugin.php]: Add DocBlock

File: Plugin/Model/Convert/OrderPlugin.php [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    )
+    {
+        $invoiceItem = $proceed($item);
+
+        $this->extensionAttributeMerger->copyAttributes($item, $invoiceItem);
+
+        return $invoiceItem;
+    }
+
+    public function aroundItemToCreditmemoItem(
```
Comment (erikhansen) [inline Plugin/Model/Convert/OrderPlugin.php]: Add DocBlock

File: Setup/UpgradeSchema.php [hunk 9]
```diff
@@ -385,6 +385,143 @@ public function upgrade(SchemaSetupInterface $setup, ModuleContextInterface $con
 
         // TODO: Add foreign key on avatax_cross_border_class.cross_border_type
```
Comment (erikhansen) [inline Setup/UpgradeSchema.php]: Remove this TODO, assuming the foreign key you added below is the one referenced here.

Comment (LordZardeck) [reply Setup/UpgradeSchema.php]: It is not. I didn't add a foreign key, i added a unique index

**Pull request merged**